File: mailmeta/analyzer.py
"""Core email analysis: parse headers, structure detection, header-only heuristics."""

import re
from email import message_from_bytes
from email.header import decode_header
from email.utils import parsedate_to_datetime, parseaddr
from datetime import datetime, timezone


def decode_mime_word(val) -> str:
    """Decode RFC2047 encoded-words (e.g. =?UTF-8?B?...?=) to plain text."""
    if not val:
        return ""
    out = []
    try:
        parts = decode_header(val)
        for chunk, enc in parts:
            if isinstance(chunk, bytes):
                out.append(chunk.decode(enc or "utf-8", errors="replace"))
            else:
                out.append(chunk)
    except Exception:
        return str(val).strip()
    return "".join(out).strip()


class EmailMeta:
    """Data structure holding all metadata extracted from an email."""

    def __init__(self):
        self.message_id = ""
        self.subject = ""
        self.from_addr = ""
        self.from_name = ""
        self.sender_addr = ""
        self.reply_to = ""
        self.return_path = ""
        self.date = ""
        self.date_ts = 0
        self.content_type = ""
        self.received_chain = []
        self.ip_address = ""
        self.envelope_from = ""
        self.headers = {}
        self.raw = b""
        self.spf = {"status": "unknown"}
        self.dkim = {"status": "unknown"}
        self.dmarc = {"status": "unknown"}
        self.authentication_results = []
        self.risk_score = 0
        self.risk_level = "unknown"
        self.findings = []
        self.ptr = None
        self.geo = None
        self.chain_ips = []
        self.chain_geo = []

    def to_dict(self, include_raw=False):
        d = {
            "message_id": self.message_id,
            "subject": self.subject,
            "from_addr": self.from_addr,
            "from_name": self.from_name,
            "sender_addr": self.sender_addr,
            "reply_to": self.reply_to,
            "return_path": self.return_path,
            "date": self.date,
            "date_ts": self.date_ts,
            "content_type": self.content_type,
            "received_chain": self.received_chain,
            "ip_address": self.ip_address,
            "envelope_from": self.envelope_from,
            "spf": self.spf,
            "dkim": self.dkim,
            "dmarc": self.dmarc,
            "authentication_results": self.authentication_results,
            "risk_score": self.risk_score,
            "risk_level": self.risk_level,
            "findings": self.findings,
            "ptr": self.ptr,
            "geo": self.geo,
            "chain_ips": self.chain_ips,
            "chain_geo": self.chain_geo,
        }
        if include_raw:
            d["raw"] = self.raw.decode("utf-8", errors="replace")
        return d


def parse_email(raw: bytes) -> EmailMeta:
    """Parse a raw eml message into an EmailMeta structure."""
    meta = EmailMeta()
    meta.raw = raw
    msg = message_from_bytes(raw)

    for key in msg.keys():
        meta.headers[key.lower()] = msg.get_all(key) or []

    # Message-ID
    mid = msg.get("Message-ID", "")
    meta.message_id = mid.strip()

    # Subject
    subj = msg.get("Subject", "")
    meta.subject = decode_mime_word(subj)

    # From
    frm = msg.get("From", "")
    fname, faddr = parseaddr(frm)
    meta.from_name = decode_mime_word(fname)
    meta.from_addr = faddr

    # Sender
    snd = msg.get("Sender", "")
    if snd:
        _, saddr = parseaddr(snd)
        meta.sender_addr = saddr

    # Reply-To
    rto = msg.get("Reply-To", "")
    if rto:
        _, raddr = parseaddr(rto)
        meta.reply_to = raddr

    # Return-Path / envelope from
    rp = msg.get("Return-Path", "")
    if rp:
        _, rpaddr = parseaddr(rp)
        meta.return_path = rpaddr
        meta.envelope_from = rpaddr

    # Date
    dt = msg.get("Date", "")
    meta.date = dt.strip()
    try:
        parsed = parsedate_to_datetime(dt)
        meta.date_ts = int(parsed.timestamp())
    except Exception:
        meta.date_ts = 0

    # Content-Type
    ct = msg.get("Content-Type", "")
    meta.content_type = ct.strip()

    # Received chain
    received = msg.get_all("Received")
    if received:
        meta.received_chain = [r.strip() for r in received]

    # Extract IP from Received chain (origin = last Received header typically)
    meta.ip_address = extract_recipient_ip(meta.received_chain)

    # Authentication-Results
    auth_res = msg.get_all("Authentication-Results")
    if auth_res:
        meta.authentication_results = [r.strip() for r in auth_res]

    return meta


IP4_RE = re.compile(r"\[(?:IPv6:)?\]|(?:\d{1,3}\.){3}\d{1,3}|\[IPv6:[0-9a-fA-F:]+\]")
IPV4 = re.compile(r"(?<!\d)(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})(?!\d)")
IPV6 = re.compile(r"\[(?:IPv6:)?([0-9a-fA-F:]{2,39})\]")


def extract_chain_ips(received_chain) -> list[str]:
    """All valid IPs found across the Received chain, useful for hop tracing."""
    out = []
    if not received_chain:
        return out
    for header in received_chain:
        for m in IPV4.finditer(header):
            a, b, c, d = m.groups()
            if all(0 <= int(x) <= 255 for x in (a, b, c, d)):
                ip = m.group(0)
                if ip not in out:
                    out.append(ip)
        for m in IPV6.finditer(header):
            ip = m.group(1)
            if ip not in out:
                out.append(ip)
    return out


def extract_recipient_ip(received_chain) -> str:
    """Extract the origin IP. In a Received chain, headers appear newest-first,
    so the LAST Received header is the origin/edge hop (sender's IP)."""
    origin_ip = ""
    if not received_chain:
        return origin_ip
    # Iterate from last (oldest) header first - that's closest to sender
    for header in reversed(received_chain):
        candidates = []
        for m in IPV4.finditer(header):
            a, b, c, d = m.groups()
            if all(0 <= int(x) <= 255 for x in (a, b, c, d)):
                candidates.append(m.group(0))
        if not candidates:
            for m in IPV6.finditer(header):
                candidates.append(m.group(1))
        if candidates:
            origin_ip = candidates[0]
            break
    return origin_ip

File: src/analyzer.ts
/**
 * Core email header analysis: RFC2047 decoding, MIME header extraction, IP detection.
 */

export interface EmailMeta {
  message_id: string;
  subject: string;
  from_addr: string;
  from_name: string;
  sender_addr: string;
  reply_to: string;
  return_path: string;
  envelope_from: string;
  date: string;
  date_ts: number;
  content_type: string;
  received_chain: string[];
  ip_address: string;
  headers: Record<string, string[]>;
  raw: string;
  spf: {
    status: string;
    explanation?: string;
    spf_record?: string | null;
    mechanisms?: Array<{ mechanism: string; value: string; qualifier: string; matched: boolean; hit: boolean }>;
  };
  dkim: {
    status: string;
    detail?: string;
    signature_domain?: string | null;
  };
  dmarc: {
    status: string;
    policy?: string | null;
    sp?: string | null;
    pct?: string | null;
    detail?: string;
    record?: string | null;
  };
  authentication_results: string[];
  risk_score: number;
  risk_level: string;
  findings: string[];
  ptr?: string | null;
  geo?: {
    status: string;
    ip?: string;
    country?: string;
    countryCode?: string;
    region?: string;
    regionName?: string;
    city?: string;
    lat?: number;
    lon?: number;
    timezone?: string;
    isp?: string;
    org?: string;
    note?: string;
    error?: string;
  } | null;
  chain_ips: string[];
}

export function decodeMimeWord(val: string): string {
  if (!val) return "";
  // Decode RFC 2047 words: =?charset?encoding?encoded_text?=
  const regex = /=\?([^?]+)\?([BQbq])\?([^?]+)\?=/g;
  return val.replace(regex, (_match, charset, encoding, text) => {
    try {
      const enc = encoding.toUpperCase();
      if (enc === "B") {
        return Buffer.from(text, "base64").toString(charset.toLowerCase() === "utf-8" ? "utf-8" : "latin1");
      } else if (enc === "Q") {
        // Quoted-Printable decode
        const unescaped = text.replace(/_/g, " ").replace(/=([0-9A-Fa-f]{2})/g, (_m: string, hex: string) =>
          String.fromCharCode(parseInt(hex, 16))
        );
        return unescaped;
      }
    } catch {
      return text;
    }
    return text;
  }).trim();
}

export function parseAddr(input: string): { name: string; address: string } {
  if (!input) return { name: "", address: "" };
  const trimmed = input.trim();
  const angleMatch = trimmed.match(/^(.*?)\s*<([^>]+)>\s*$/);
  if (angleMatch) {
    let name = angleMatch[1].trim();
    if ((name.startsWith('"') && name.endsWith('"')) || (name.startsWith("'") && name.endsWith("'"))) {
      name = name.slice(1, -1).trim();
    }
    return {
      name: decodeMimeWord(name),
      address: angleMatch[2].trim().toLowerCase(),
    };
  }

  if (trimmed.includes("@")) {
    const clean = trimmed.replace(/[<>]/g, "").trim();
    return { name: "", address: clean.toLowerCase() };
  }

  return { name: decodeMimeWord(trimmed), address: "" };
}

const IPV4_REGEX = /(?<!\d)(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})(?!\d)/g;
const IPV6_REGEX = /\[(?:IPv6:)?([0-9a-fA-F:]{2,39})\]/g;

export function extractChainIps(receivedChain: string[]): string[] {
  const out: string[] = [];
  if (!receivedChain || !receivedChain.length) return out;

  for (const header of receivedChain) {
    let match: RegExpExecArray | null;
    const v4Copy = new RegExp(IPV4_REGEX);
    while ((match = v4Copy.exec(header)) !== null) {
      const [, a, b, c, d] = match;
      if ([a, b, c, d].every((x) => Number(x) >= 0 && Number(x) <= 255)) {
        const ip = match[0];
        if (!out.includes(ip)) out.push(ip);
      }
    }
    const v6Copy = new RegExp(IPV6_REGEX);
    while ((match = v6Copy.exec(header)) !== null) {
      const ip = match[1];
      if (!out.includes(ip)) out.push(ip);
    }
  }
  return out;
}

export function extractRecipientIp(receivedChain: string[]): string {
  if (!receivedChain || !receivedChain.length) return "";
  // In Received chain, headers are listed newest-first. The LAST header is closest to the sender.
  const reversed = [...receivedChain].reverse();
  for (const header of reversed) {
    const candidates: string[] = [];
    const v4Copy = new RegExp(IPV4_REGEX);
    let match: RegExpExecArray | null;
    while ((match = v4Copy.exec(header)) !== null) {
      const [, a, b, c, d] = match;
      if ([a, b, c, d].every((x) => Number(x) >= 0 && Number(x) <= 255)) {
        candidates.push(match[0]);
      }
    }
    if (!candidates.length) {
      const v6Copy = new RegExp(IPV6_REGEX);
      while ((match = v6Copy.exec(header)) !== null) {
        candidates.push(match[1]);
      }
    }
    if (candidates.length > 0) {
      return candidates[0];
    }
  }
  return "";
}

/**
 * Parse raw RFC822 / EML content into headers and structure
 */
export function parseEmail(rawContent: string): EmailMeta {
  const normalized = rawContent.replace(/\r\n/g, "\n");
  const headersEnd = normalized.indexOf("\n\n");
  const headerSection = headersEnd !== -1 ? normalized.slice(0, headersEnd) : normalized;

  const headerLines = headerSection.split("\n");
  const foldedHeaders: string[] = [];

  for (const line of headerLines) {
    if ((line.startsWith(" ") || line.startsWith("\t")) && foldedHeaders.length > 0) {
      foldedHeaders[foldedHeaders.length - 1] += " " + line.trim();
    } else if (line.trim().length > 0) {
      foldedHeaders.push(line);
    }
  }

  const headers: Record<string, string[]> = {};
  for (const h of foldedHeaders) {
    const colonIdx = h.indexOf(":");
    if (colonIdx !== -1) {
      const key = h.slice(0, colonIdx).trim().toLowerCase();
      const val = h.slice(colonIdx + 1).trim();
      if (!headers[key]) headers[key] = [];
      headers[key].push(val);
    }
  }

  const getFirst = (key: string): string => (headers[key.toLowerCase()] ? headers[key.toLowerCase()][0] : "");

  const mid = getFirst("message-id");
  const subj = decodeMimeWord(getFirst("subject"));
  const fromRaw = getFirst("from");
  const { name: fromName, address: fromAddr } = parseAddr(fromRaw);

  const senderRaw = getFirst("sender");
  const { address: senderAddr } = parseAddr(senderRaw);

  const replyToRaw = getFirst("reply-to");
  const { address: replyTo } = parseAddr(replyToRaw);

  const returnPathRaw = getFirst("return-path");
  const { address: returnPath } = parseAddr(returnPathRaw);

  const dateStr = getFirst("date");
  let dateTs = 0;
  if (dateStr) {
    const parsed = Date.parse(dateStr);
    if (!isNaN(parsed)) {
      dateTs = Math.floor(parsed / 1000);
    }
  }
  if (!dateTs) {
    dateTs = Math.floor(Date.now() / 1000);
  }

  const contentType = getFirst("content-type");
  const receivedChain = headers["received"] || [];
  const authResults = [
    ...(headers["authentication-results"] || []),
    ...(headers["arc-authentication-results"] || []),
    ...(headers["received-spf"] || []),
    ...(headers["x-google-dkim-signature"] || []),
  ];
  const originIp = extractRecipientIp(receivedChain);
  const chainIps = extractChainIps(receivedChain);

  const meta: EmailMeta = {
    message_id: mid,
    subject: subj,
    from_addr: fromAddr,
    from_name: fromName,
    sender_addr: senderAddr,
    reply_to: replyTo,
    return_path: returnPath,
    envelope_from: returnPath || fromAddr,
    date: dateStr || new Date().toUTCString(),
    date_ts: dateTs,
    content_type: contentType,
    received_chain: receivedChain,
    ip_address: originIp,
    headers: headers,
    raw: rawContent,
    spf: { status: "unknown" },
    dkim: { status: "unknown" },
    dmarc: { status: "unknown" },
    authentication_results: authResults,
    risk_score: 0,
    risk_level: "unknown",
    findings: [],
    ptr: null,
    geo: null,
    chain_ips: chainIps,
  };

  return meta;
}
